### PR TITLE
uSSL:  Add compile option for a finaliser

### PR DESCRIPTION
**Pull request by MrSurly**
Per https://github.com/micropython/micropython-esp32/issues/209#issuecomment-339855157

### CODE REVIEW

**Comment by MrSurly** [comment]
If accepted, please propagate to ESP32.

**Comment by dpgeorge** [comment]
A problem that this brings up is that the GC is not reentrant.  That means a finaliser cannot explicitly free memory on the uPy heap.  In the best case a call to gc_free() in the finaliser does nothing.  In the worst case (eg with threading enabled) it leads to deadlocks.  That hasn't been an issue in the past because all uses of finalisers are to close files/sockets, not to free memory on the uPy heap (which is anyway done automatically).

The concern then is that for ports that use the uPy heap for allocating mbedtls objects (which is not the case for esp32 which has a separate heap), this patch will lead to bugs.

**Review by pfalcon — CHANGES_REQUESTED** [review]
Please add to the commit message a detailed description why the change is made and what it brings. (This generally applies to all commits.)

**Comment by MrSurly** [comment]
@pfalcon Amended commit message

**Comment by MrSurly** [comment]
@dpgeorge Added a guard `#ifdef` around finaliser.

**Review by pfalcon — APPROVED** [review]
Thanks.

**Comment by pfalcon** [comment]
Two commits can be squashed into one btw.

**Review by pfalcon — CHANGES_REQUESTED** [review]
Oh, btw, all options added should be listed in py/mpconfig.h, or we would have lost track of them long ago.

**Comment by MrSurly** [comment]
@pfalcon 
* Updated py/mpconfig.h
* Generalized the define to `MICROPY_PY_USSL_FINALISER`
* Added this define, with a comment to `py/mpconfig.h`
* Squashed into one commit

File: py/mpconfig.h [hunk 1]
```diff
@@ -1109,6 +1109,8 @@ typedef double mp_float_t;
 
 #ifndef MICROPY_PY_USSL
 #define MICROPY_PY_USSL (0)
+// Whether to add finaliser code to uSSL
```
Comment (pfalcon) [inline py/mpconfig.h]: Better to say "Whether to add finaliser to ussl objects".

**Review by pfalcon — APPROVED** [review]
Looks good, thanks. Minor documenting comment suggestion.

**Comment by dpgeorge** [comment]
Merged in 74ec52d85758ad9da7a3abb24257511b22d74964

**Pull request closed without merging**